File: cybernovels/cli/go/converter/converter.go
package converter

import (
	"context"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"

	"github.com/chromedp/chromedp"
	"github.com/yuin/goldmark"
)

// ConvertFile selects the conversion path based on file extensions.
func ConvertFile(inputPath, outputPath string) error {
	inInExt := filepath.Ext(inputPath)
	outExt := filepath.Ext(outputPath)

	fmt.Printf("Native conversion from '%s' to '%s'\n", inExt, outExt)

	switch {
	case inExt == ".md" && outExt == ".html":
		return convertMdToHTML(inputPath, outputPath)
	case in_ext == ".html" && out_ext == ".pdf":
		return convertHTMLToPDF(inputPath, outputPath)
	case inExt == ".md" && outExt == ".pdf":
		// Chain: md -> html -> pdf
		htmlTempPath := inputPath[0:len(inputPath)-len(inExt)] + ".temp.html"
		defer os.Remove(htmlTempPath)

		fmt.Println("Step 1: md -> html")
		if err := convertMdToHTML(inputPath, htmlTempPath); err != nil {
			return err
		}
		fmt.Println("Step 2: html -> pdf")
		return convertHTMLToPDF(htmlTempPath, outputPath)
	default:
		return fmt.Errorf("native conversion from %s to %s is not supported", inExt, outExt)
	}
}

func convertMdToHTML(inPath, outPath string) error {
	source, err := ioutil.ReadFile(inPath)
	if err != nil {
		return err
	}
	outFile, err := os.Create(outPath)
	if err != nil {
		return err
	}
	defer outFile.Close()

	return goldmark.Convert(source, outFile)
}

func convertHTMLToPDF(inPath, outPath string) error {
	// We need the absolute path for chromedp
	absInPath, err := filepath.Abs(inPath)
	if err != nil {
		return err
	}

	ctx, cancel := chromedp.NewContext(context.Background())
	defer cancel()

	var buf []byte
	err = chromedp.Run(ctx,
		chromedp.Navigate("file://"+absInPath),
		chromedp.ActionFunc(func(ctx context.Context) error {
			var err error
			buf, _, err = page.PrintToPDF().Do(ctx)
			return err
		}),
	)
	if err != nil {
		return err
	}

	return ioutil.WriteFile(outPath, buf, 0644)
}

File: cybernovels/cli/python/doconvrt_native/converter.py
"""
Native conversion library for Python.
Uses markdown-it-py for Markdown -> HTML, and WeasyPrint for HTML -> PDF.
"""
from markdown_it import MarkdownIt
from weasyprint import HTML
import os

def convert_md_to_html(in_path, out_path):
    """Converts Markdown to HTML."""
    md = MarkdownIt()
    with open(in_path, 'r', encoding='utf-8') as f:
        text = f.read()
    html_content = md.render(text)
    with open(out_path, 'w', encoding='utf-8') as f:
        f.write(html_content)

def convert_html_to_pdf(in_path, out_path):
    """Converts HTML to PDF."""
    HTML(in_path).write_pdf(out_path)

def convert_file(input_path, output_path):
    """
    Selects the correct conversion path based on file extensions.
    """
    in_ext = os.path.splitext(input_path)[1].lower()
    out_ext = os.path.splitext(output_path)[1].lower()

    print(f"Native conversion from '{in_ext}' to '{out_ext}'")

    if in_ext == '.md' and out_ext == '.html':
        convert_md_to_html(input_path, output_path)
    elif in_ext == '.html' and out_ext == '.pdf':
        convert_html_to_pdf(input_path, output_path)
    elif in_ext == '.md' and out_ext == '.pdf':
        # Chain conversion: md -> html -> pdf
        html_temp_path = os.path.splitext(input_path)[0] + ".temp.html"
        try:
            print("Step 1: md -> html")
            convert_md_to_html(input_path, html_temp_path)
            print("Step 2: html -> pdf")
            convert_html_to_pdf(html_temp_path, output_path)
        finally:
            # Clean up temporary file
            if os.path.exists(html_temp_path):
                os.remove(html_temp_path)
    else:
        raise NotImplementedError(f"Native conversion from {in_ext} to {out_ext} is not supported.")
